### PR TITLE
Rename, document, update joins/set operations

### PR DESCRIPTION
**Pull request by harendra-kumar**

### CODE REVIEW

File: core/src/Streamly/Internal/Data/Stream/Container.hs [hunk 1]
```diff
@@ -47,12 +48,14 @@ import qualified Streamly.Internal.Data.Stream.Transformer as Stream
...
-nub :: (Monad m, Ord a) => Stream m a -> Stream m a
-nub (Stream step1 state1) = Stream step (Set.empty, state1)
+-- | Return a subsequence of the stream removing any duplicate elements.
+--
+-- The memory used is proportional to the number of unique elements in the
+-- stream. One way to limit the memory is to  use @take@ on the resulting
+-- stream to limit the unique elements in the stream.
+{-# INLINE_NORMAL ordNub #-}
+ordNub :: (Monad m, Ord a) => Stream m a -> Stream m a
+ordNub (Stream step1 state1) = Stream step (Set.empty, state1)
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Stream/Container.hs:58]: Why ord prefix? Does it return an ordered stream or require an ordered stream? Can that be mentioned in the documentation?

File: core/src/Streamly/Internal/Data/Stream/Container.hs [hunk 2]
```diff
@@ -90,10 +93,10 @@ toMap =
 -- Time: O(m + n)
 --
 -- /Pre-release/
-{-# INLINE joinInner #-}
-joinInner :: (Monad m, Ord k) =>
+{-# INLINE innerOrdJoin #-}
+innerOrdJoin :: (Monad m, Ord k) =>
     Stream m (k, a) -> Stream m (k, b) -> Stream m (k, a, b)
-joinInner s1 s2 =
+innerOrdJoin s1 s2 =
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Stream/Container.hs:99]: Same comment as the above. Significance or ord?

File: core/src/Streamly/Internal/Data/Stream/Container.hs [hunk 3]
```diff
@@ -152,19 +156,17 @@ joinLeftGeneric eq s1 s2 = Stream.evalStateT (return False) $ unCross $ do
...
 -- Time: O(m + n)
 --
 -- /Pre-release/
-{-# INLINE joinLeft #-}
-joinLeft :: (Ord k, Monad m) =>
+{-# INLINE leftOrdJoin #-}
+leftOrdJoin :: (Ord k, Monad m) =>
     Stream m (k, a) -> Stream m (k, b) -> Stream m (k, a, Maybe b)
-joinLeft s1 s2 =
+leftOrdJoin s1 s2 =
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Stream/Container.hs:169]: `ord`?

File: core/src/Streamly/Internal/Data/Stream/Container.hs [hunk 4]
```diff
@@ -106,32 +109,33 @@ joinInner s1 s2 =
...
+-- | Like 'innerJoin' but emits @(a, Just b)@ whenever a and b are equal, for
+-- those @a@'s that are not equal to any @b@ emits @(a, Nothing)@.
 --
--- The second stream is evaluated multiple times. If the stream is a
--- consume-once stream then the caller should cache it in an 'Data.Array.Array'
--- before calling this function. Caching may also improve performance if the
--- stream is expensive to evaluate.
+-- This is a generalization of 'innerJoin' to include all elements from the
+-- left stream and not just those which have an equal in the right stream. This
+-- is not a commutative operation, the order of the stream arguments matters.
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Stream/Container.hs:123]:
```suggestion
-- is not a commutative operation, the order of the stream arguments matter.
```

File: core/src/Streamly/Internal/Data/Stream/Container.hs [hunk 5]
```diff
@@ -254,18 +261,18 @@ joinOuterGeneric eq s1 s =
...
 --
 -- Space: O(m + n)
 --
 -- Time: O(m + n)
 --
 -- /Pre-release/
-{-# INLINE joinOuter #-}
-joinOuter ::
+{-# INLINE outerOrdJoin #-}
+outerOrdJoin ::
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Stream/Container.hs:272]: I'm assuming the output is ordered.

File: core/src/Streamly/Internal/Data/Stream/Top.hs [hunk 6]
```diff
@@ -110,29 +145,41 @@ strideFromThen offset stride =
...
+-- before calling this function. Caching may also improve performance if the
+-- stream is expensive to evaluate.
 --
--- You should almost always prefer @joinInnerOrd@ over 'joinInnerGeneric' if
--- possible. @joinInnerOrd@ is an order of magnitude faster but may take more
--- space for caching the second stream.
+-- If you care about performance this function should be your last choice among
+-- all inner joins. 'Streamly.Internal.Data.Unfold.innerJoin' is a much faster
+-- fused alternative. 'sortedInnerJoin' is a faster alternative when streams
+-- are sorted. If the type has 'Ord' instance then 'ordInnerJoin' is an order
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Stream/Top.hs:166]:
```suggestion
-- are sorted. If the type has 'Ord' instance then 'innerOrdJoin' is an order
```

File: core/src/Streamly/Internal/Data/Stream/Nesting.hs [hunk 7]
```diff
@@ -593,49 +592,6 @@ mergeFstBy :: -- Monad m =>
...
 
--------------------------------------------------------------------------------
--- Intersection of sorted streams
--------------------------------------------------------------------------------
-
--- Assuming the streams are sorted in ascending order
-{-# INLINE_NORMAL intersectBySorted #-}
-intersectBySorted :: Monad m
-    => (a -> a -> Ordering) -> Stream m a -> Stream m a -> Stream m a
-intersectBySorted cmp (Stream stepa ta) (Stream stepb tb) =
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Stream/Nesting.hs:604]: Where is this moved to? Is this not `sortedInnerJoin`?

Comment (adithyaov) [reply core/src/Streamly/Internal/Data/Stream/Nesting.hs:604]: Ah, a more generic version tho.

File: core/src/Streamly/Internal/Data/Stream/Top.hs [hunk 8]
```diff
@@ -206,129 +253,196 @@ filterStreamWith :: Monad m =>
 filterStreamWith fld member s1 s2 =
     Stream.concatEffect
         $ do
-            xs <- Stream.fold fld s1
-            return $ Stream.filter (`member` xs) s2
+            xs <- Stream.fold fld s2
+            return $ Stream.filter (`member` xs) s1
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Stream/Top.hs:257]: Is this quadratic?
Can we use a set here?

File: core/src/Streamly/Internal/Data/Stream/Top.hs [hunk 9]
```diff
@@ -206,129 +253,196 @@ filterStreamWith :: Monad m =>
...
 -- Time: O(m+n)
 --
 -- /Pre-release/
-{-# INLINE filterInStreamAscBy #-}
-filterInStreamAscBy :: Monad m =>
+{-# INLINE_NORMAL sortedIntersectBy #-}
+sortedIntersectBy :: Monad m =>
     (a -> a -> Ordering) -> Stream m a -> Stream m a -> Stream m a
-filterInStreamAscBy eq s1 s2 = Stream.intersectBySorted eq s2 s1
+sortedIntersectBy cmp (Stream stepa ta) (Stream stepb tb) =
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Stream/Top.hs:318]: Ah, it's moved here.

**Review by adithyaov — CHANGES_REQUESTED** [review]

File: core/src/Streamly/Internal/Data/Stream/Container.hs (hunk 4 above)
Comment (harendra-kumar) [reply core/src/Streamly/Internal/Data/Stream/Container.hs:123]: original is correct - `the order matters`.

File: core/src/Streamly/Internal/Data/Stream/Top.hs (hunk 8 above)
Comment (harendra-kumar) [reply core/src/Streamly/Internal/Data/Stream/Top.hs:257]: This is for unconstrained version. The version with Set would be separate with an Ord constraint.

**Pull request merged**